adapters/Vungle/VungleAdapter/Bidding/GADMediationVungleBanner.m

```diff
@@ -101,154 +98,133 @@ - (void)requestBannerAd {
 }
 
 - (void)loadAd {
-  NSError *error = [[GADMAdapterVungleBiddingRouter sharedInstance] loadAdWithDelegate:self];
-  if (error) {
-    _adLoadCompletionHandler(nil, error);
+  _bannerAd = [[VungleBanner alloc] initWithPlacementId:self.desiredPlacement
+                                                   size:[self convertGADAdSizeToBannerSize]];
+  _bannerAd.delegate = self;
+  VungleAdsExtras *extras = [[VungleAdsExtras alloc] init];
+  [extras setWithWatermark:[_adConfiguration.watermark base64EncodedStringWithOptions:0]];
+  [_bannerAd setWithExtras:extras];
+  [_bannerAd load:_adConfiguration.bidResponse];
+}
+
+- (GADAdSize)vungleAdSizeForAdSize:(GADAdSize)adSize {
+  // It has to match for MREC, otherwise it would be a banner with flexible size
+  if (adSize.size.height == GADAdSizeMediumRectangle.size.height &&
+      adSize.size.width == GADAdSizeMediumRectangle.size.width) {
+    return GADAdSizeMediumRectangle;
   }
-}
 
-- (NSError *)renderAd {
-  VungleAdNetworkExtras *extras = (VungleAdNetworkExtras *)[_adConfiguration extras];
-  NSMutableDictionary *options = nil;
-  if (extras) {
-    options = [[NSMutableDictionary alloc] init];
-    if (extras.muteIsSet) {
-      GADMAdapterVungleMutableDictionarySetObjectForKey(options, VunglePlayAdOptionKeyStartMuted,
-                                                        @(extras.muted));
-    }
-    if (extras.userId) {
-      GADMAdapterVungleMutableDictionarySetObjectForKey(options, VunglePlayAdOptionKeyUser,
-                                                        extras.userId);
-    }
-    if (extras.flexViewAutoDismissSeconds) {
-      GADMAdapterVungleMutableDictionarySetObjectForKey(
-          options, VunglePlayAdOptionKeyFlexViewAutoDismissSeconds,
-          @(extras.flexViewAutoDismissSeconds));
+  // An array of supported ad sizes.
+  GADAdSize shortBannerSize = GADAdSizeFromCGSize(kVNGBannerShortSize);
+  NSArray<NSValue *> *potentials = @[
+    NSValueFromGADAdSize(GADAdSizeBanner), NSValueFromGADAdSize(GADAdSizeLeaderboard),
+    NSValueFromGADAdSize(shortBannerSize)
+  ];
+
+  GADAdSize closestSize = GADClosestValidSizeForAdSizes(adSize, potentials);
+  CGSize size = CGSizeFromGADAdSize(closestSize);
+  if (size.height == GADAdSizeBanner.size.height) {
+    if (size.width < GADAdSizeBanner.size.width) {
+      return shortBannerSize;
+    } else {
+      return GADAdSizeBanner;
     }
+  } else if (size.height == GADAdSizeLeaderboard.size.height) {
+    return GADAdSizeLeaderboard;
   }
-  NSError *bannerError = nil;
-  [VungleSDK.sharedSDK addAdViewToView:_bannerView
-                           withOptions:options
-                           placementID:self.desiredPlacement
-                              adMarkup:[self bidResponse]
-                                 error:&bannerError];
-  return bannerError;
+  return GADAdSizeInvalid;
 }
 
-- (void)cleanUp {
-  if (_didBannerFinishPresenting) {
-    return;
+- (BannerSize)convertGADAdSizeToBannerSize {
```

Comment:
Yes. It can.
The Vungle SDK doesn't have an "invalid" in the banner size enum so it would default to our regular size banner and then would fail presentation because of a size mismatch instead of failing the load request.

Before I change it, can I get a thought about whether it is better to fail on the load vs the playback?

Reply:
Moved this method to utils to follow review from joshh.